### PR TITLE
Support for hybrid CPU

### PR DESCRIPTION
**Pull request by X0rg**
Hello @anrieff,

Good new: here are major changes to support hybrid CPUs like Intel Alder Lake!
It is challenging to keep compatibility with existing tests and support new tests with multiples RAW and different CPU types, but I found a way to achieve that goal.

I cleaned up my commits a bit on this branch (I do not guarantee individual commit build successfully due to rebase and cherry-picking, but current branch status is good).

`cpuid_tool` output for Intel Core i912900K is looking like that:
```
CPUID is present
CPU Info for type #0:
------------------
  arch       : x86
  purpose    : performance
  vendor_str : `GenuineIntel'
  vendor id  : 0
  brand_str  : `12th Gen Intel(R) Core(TM) i9-12900K'
  family     : 6 (06h)
  model      : 7 (07h)
  stepping   : 2 (02h)
  ext_family : 6 (06h)
  ext_model  : 151 (97h)
  num_cores  : 8
  num_logical: 16
  tot_logical: 24
  affi_mask  : 0x0000FFFF
  L1 D cache : 48 KB
  L1 I cache : 32 KB
  L2 cache   : 1280 KB
  L3 cache   : 30720 KB
  L4 cache   : -1 KB
  L1D assoc. : 12-way
  L1I assoc. : 8-way
  L2 assoc.  : 10-way
  L3 assoc.  : 12-way
  L4 assoc.  : -1-way
  L1D line sz: 64 bytes
  L1I line sz: 64 bytes
  L2 line sz : 64 bytes
  L3 line sz : 64 bytes
  L4 line sz : -1 bytes
  SSE units  : 128 bits (non-authoritative)
  code name  : `Alder Lake-S (Core i9)'
  features   : fpu vme de pse tsc msr pae mce cx8 apic mtrr sep pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe pni pclmul dts64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm sse4_1 sse4_2 xd movbe popcnt aes xsave osxsave avx rdtscp lm lahf_lm abm constant_tsc fma3 f16c rdrand x2apic avx2 bmi1 bmi2 sha_ni rdseed adx
CPU Info for type #1:
------------------
  arch       : x86
  purpose    : efficiency
  vendor_str : `GenuineIntel'
  vendor id  : 0
  brand_str  : `12th Gen Intel(R) Core(TM) i9-12900K'
  family     : 6 (06h)
  model      : 7 (07h)
  stepping   : 2 (02h)
  ext_family : 6 (06h)
  ext_model  : 151 (97h)
  num_cores  : 8
  num_logical: 8
  tot_logical: 24
  affi_mask  : 0x00FF0000
  L1 D cache : 32 KB
  L1 I cache : 64 KB
  L2 cache   : 2048 KB
  L3 cache   : 30720 KB
  L4 cache   : -1 KB
  L1D assoc. : 8-way
  L1I assoc. : 8-way
  L2 assoc.  : 16-way
  L3 assoc.  : 12-way
  L4 assoc.  : -1-way
  L1D line sz: 64 bytes
  L1I line sz: 64 bytes
  L2 line sz : 64 bytes
  L3 line sz : 64 bytes
  L4 line sz : -1 bytes
  SSE units  : 128 bits (non-authoritative)
  code name  : `Alder Lake-S (Core i9)'
  features   : fpu vme de pse tsc msr pae mce cx8 apic mtrr sep pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe pni pclmul dts64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm sse4_1 sse4_2 xd movbe popcnt aes xsave osxsave avx rdtscp lm lahf_lm abm constant_tsc fma3 f16c rdrand x2apic avx2 bmi1 bmi2 sha_ni rdseed adx
```
We can clearly identify 8 performance cores with HT and 8 efficiency cores without HT. So 16 + 8 threads, total logical CPU count is 24.

For non-hybrid CPUs, I decided to create a "general" purpose, because I do not think the concept of "performance" or "efficiency" makes sense for such case. Example for my AMD Ryzen 3600X:
```
CPUID is present
CPU Info for type #0:
------------------
  arch       : x86
  purpose    : general
  vendor_str : `AuthenticAMD'
  vendor id  : 1
  brand_str  : `AMD Ryzen 5 3600X 6-Core Processor             '
  family     : 15 (0Fh)
  model      : 1 (01h)
  stepping   : 0 (00h)
  ext_family : 23 (17h)
  ext_model  : 113 (71h)
  num_cores  : 6
  num_logical: 12
  tot_logical: 12
  affi_mask  : 0x00000FFF
  L1 D cache : 32 KB
  L1 I cache : 32 KB
  L2 cache   : 512 KB
  L3 cache   : 32768 KB
  L4 cache   : -1 KB
  L1D assoc. : 8-way
  L1I assoc. : 8-way
  L2 assoc.  : 8-way
  L3 assoc.  : 16-way
  L4 assoc.  : -1-way
  L1D line sz: 64 bytes
  L1I line sz: 64 bytes
  L2 line sz : 64 bytes
  L3 line sz : 64 bytes
  L4 line sz : -1 bytes
  SSE units  : 256 bits (authoritative)
  code name  : `Ryzen 5 (Matisse)'
  features   : fpu vme de pse tsc msr pae mce cx8 apic mtrr sep pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht pni pclmul monitor ssse3 cx16 sse4_1 sse4_2 syscall movbe popcnt aes xsave osxsave avx mmxext nx fxsr_opt rdtscp lm lahf_lm cmp_legacy svm abm misalignsse sse4a 3dnowprefetch osvw ibs skinit wdt ts ttp tm_amd hwpstate constant_tsc fma3 f16c rdrand cpb aperfmperf avx2 bmi1 bmi2 sha_ni rdseed adx
```

CPU type delimiters for `cpuid_tool` output and tests are arbitrary, feel free to propose something else.

In 4f80964db5bab011c1893ad17adcaec20d6d7fae, I wrote the `set_cpu_affinity()` functions based on examples I found on the Internet. I tested this function on my Linux system but I did not test it on other OS (CI with macOS and Windows build fine, but it does not reflect the behavior of real life systems).

Please let me know if you have time to review this PR. It introduces a backwards-incompatible change, so I think the next release will be 0.6.0.

### CODE REVIEW

**Comment by anrieff** [comment]
I'll review it.
Great work as always - this was a huge task, and I'm glad you found time for it!

**Review by anrieff — DISMISSED** [review]

File: libcpuid/CMakeLists.txt [hunk 1]
```diff
@@ -8,8 +8,11 @@ set(cpuid_sources
     asm-bits.c)
 
 if(WIN32)
-   list(APPEND cpuid_sources msrdriver.c)
+  list(APPEND cpuid_sources msrdriver.c)
 endif()
+if(UNIX)
+  add_compile_definitions(_GNU_SOURCE)
```
Comment (anrieff) [inline libcpuid/CMakeLists.txt:14]: Need to port this for autotools, e.g. add

```
if test "$build_windows" = "no"; then
	AM_CPPFLAGS="$AM_CPPFLAGS -D_GNU_SOURCE"
fi
```

before `AM_CONDITIONAL([WINDOWS], [test "$build_windows" = "yes"])` in `autoconf.ac`

File: libcpuid/cpuid_main.c [hunk 2]
```diff
@@ -682,8 +682,77 @@ int cpu_ident_internal(struct cpu_raw_data_t* raw, struct cpu_id_t* data, struct
...
+int cpu_identify_all(struct cpu_raw_data_array_t *raw_array, struct system_id_t* system)
+{
+	int cur_error = set_error(ERR_OK);
+	int ret_error = set_error(ERR_OK);	bool is_new_cpu_type;
+	uint8_t logical_cpu;
+	uint8_t cpu_type_index = 0;
+	int32_t core_previous_id = -1;
+	int32_t num_cores = 1;
+	int32_t num_logical_cpus = 1;
+	uint32_t affinity_mask = 0x00000001;
```
Comment (anrieff) [inline libcpuid/cpuid_main.c:813]: The affinity mask should be uint64_t if we want to support more than 32-thread CPUs.

**Comment by anrieff** [comment]
OK, I've grepped diagonally through all changes, and first of all - **great work**! I see you really pulled huge task without cutting any corners. Thank you for your persistence!

Then I think the current proposal has some inconsistencies in the ambition to support huge CPUs like the Ryzen 3990x (64C/128T) versus how is this ambition implemented.

E.g. `CPU_RAW_MAX` is 512 (all fine and noble - let's push the boundaries!).
However, Windows will not let you set any thread past the 64th with `SetProcessAffinityMask`.
Same with `cpu_id_t::affinity_mask`. Even if you increase it to uint64_t (as I suggested in comments), you still won't capture the full picture on a 3990x (of course, that's not a hybrid CPU, but we can expect that we'll eventually hit that limit with a hybrid CPU).

What we can do about it:
1. Don't use plain integral types for affinity masks; C does not support large enough types out of the box even for the current CPU models. Instead, use an opaque struct just like the Linux affinity API works like (using `cpu_set_t` and `CPU_ZERO`, `CPU_SET` macros, etc). You may name this structure `cpu_affinity_mask_t` or `cpu_bitset_t`. The actual opaque struct will be of course defined in terms of CPU_RAW_MAX/32 amount of uint32_ts; and you'll need to add those test/set/reset/zero macros.
2. Look into the underlying OS API limitations for setting the affinity. If an API is inherently limited to N CPUs (as is `SetProcessAffinityMask`, where N = 64), then trying to set the affinity to CPU number N (0-based) will result in one of 3 things:
    1. The API call failing, in which case we're good (we won't do full topology enumeration, but at least we don't hang)
    2. The process ends up with zero affinity mask and hangs
    3. The API call returns OK, but does not change anything. In this case we'll do incorrect enumeration.
When we run into such issues, we should have our `set_cpu_affinity()` function return 0, as if the underlying API told us "no more CPUs".

## Worries and inefficiencies

1. sizeof(cpu_raw_data_array_t) is more than 700 kbytes. The libcpuid usage examples typically show these structures being allocated on the stack, and allocating a 700kB structure on-stack is certainly a bad idea. We should either warn about this in the comments/documentation very explicitly (that we suggest allocating these structures statically or with malloc), or the `cpu_raw_data_array_t::raw` should be changed to `cpu_raw_data_t*` and the `cpu*_all_*` functions should `malloc()`/`realloc()` the memory as required.
2. In similar vein, `cpuid_serialize_all_raw_data` will write 128 CPU entries for a Ryzen 3990x. While we're not worse than e.g. AIDA64, we can certainly do better. One relatively simple idea is to only truly write out the first CPU, and the following ones can only say what's different from Logical CPU 0, e.g. the entry for 1 may look like
    ```
    _________________ Logical CPU 1 _________________
    ...
    basic_cpuid[1]=000306c3 01100800 7fdafbbf bfebfbff
    basic_cpuid[11]=00000001 00000002 00000100 00000001
    intel_fn11[0]=00000001 00000002 00000100 00000001
    intel_fn11[1]=00000004 00000008 00000201 00000001
    intel_fn11[2]=00000000 00000000 00000002 00000001
    intel_fn11[3]=00000000 00000000 00000003 00000001
    ```
    i.e. the first line of the description is "...", indicating that a diff follows. Then we only list the lines with differences.
3. The code in `cpu_identify_all` enumerates all `cpu_raw_data_t`, doing full identification, only to look into the `cpu_id_t::purpose` field in the end. We can do better, e.g. add an internal function `cpu_ident_purpose()` to do only that, so we can only call `cpu_ident_internal` when it's truly needed (for the two different CPU types on a hybrid system). Note that `cpu_ident_internal` is not super slow, but it does run some linear searches in the matchtables. For a Ryzen 3990x we'll launch it no less than 128 times with the current code.

**Review by anrieff — CHANGES_REQUESTED** [review]
Please fix the affinity mask sizes or make them opaque structs as suggested in the PR comment.

File: libcpuid/cpuid_main.c (hunk 2 above)
Comment (X0rg) [reply libcpuid/cpuid_main.c:813]: Fixed in d204dcfd427e050df35a27440bfc98d8c6c40b5e.

Comment (X0rg) [reply libcpuid/cpuid_main.c]: Well, I will follow your comment below. You are right, even a uint64_t is still a limitation.

**Comment by X0rg** [comment]
Thank you for the feedback, that is very constructive.

> Don't use plain integral types for affinity masks [...]

As suggested, switching to an opaque structure is more work in one hand but more flexible in other hand. I will not reinvent the wheel, so I will reuse `cpu_set_t`,  `CPU_ZERO` and `CPU_SET` with different names to avoid conflict with `sched.h` on Linux.

>     1. sizeof(cpu_raw_data_array_t) is more than 700 kbytes [...]

Indeed. I was like "700 kB is not so much", but it is a waste of memory and AMD is very ambitious (EPYC Bergamo should have 128c/256t per CPU), so a static array with a fixed size will always be a limit one day or another.
However, a dynamic array means that we need to provide a function to free it, so I guess every functions that allocate memory must have a note in `libcpuid.h` to indicate that the data must be freed, correct?
I think a will add functions in `libcpuid_util.c` to allocate and free this array.

>     2. In similar vein, `cpuid_serialize_all_raw_data` will write 128 CPU entries for a Ryzen 3990x. [...]

That is not a bad idea but it seems very cosmetic for a file that is not intended to be read by a human being. It can save disk space but it adds more complexity.
What about AIDA64 dumps in that case? As you can see in 9ab188879f6f42a3b95d7b6f78085b5496198984, AIDA64 RAW dumps are added "as is" to the test file. Cleaning that file will require to modify the `create_test.py` script, meaning that a CPUID parser will be necessary inside to check for duplicate lines.
That is too much work to save some kilobytes IMHO.

>     3. The code in `cpu_identify_all` enumerates all `cpu_raw_data_t`, doing full identification, only to look into the `cpu_id_t::purpose` field in the end. [...]

For Intel hybrid CPUs, `cpu_identify_all` also counts logical and physical cores, but it can be moved to `cpu_ident_purpose` too.
I am surprised that Extended Topology Enumeration Leaf does not provide correct values with hybrid CPUs: the only difference I found is num_smt = num_core for efficiency core, but num_core is always the total of physical CPU cores on the package, no matter if core is E or P. And with V2 Extended Topology Enumeration Leaf if found same values...
I can improve that and improve `cpu_request_core_type` at the same time.

**Comment by anrieff** [comment]
> However, a dynamic array means that we need to provide a function to free it, so I guess every functions that allocate memory must have a note in libcpuid.h to indicate that the data must be freed, correct?

Yes, just like the business with `cpuid_get_cpu_list()` and `cpuid_free_cpu_list()`.

> That is not a bad idea but it seems very cosmetic for a file that is not intended to be read by a human being. It can save disk space but it adds more complexity.

You're likely correct, I'm probably seeing it as easier than it actually is. Let's keep the verbose format for now, I may play around with the diff-format idea once this PR is merged.

> What about AIDA64 dumps in that case? As you can see in https://github.com/anrieff/libcpuid/commit/9ab188879f6f42a3b95d7b6f78085b5496198984, AIDA64 RAW dumps are added "as is" to the test file. Cleaning that file will require to modify the create_test.py script, meaning that a CPUID parser will be necessary inside to check for duplicate lines.

This can be solved quite elegantly actually - if `cpuid_tool` can read all four input formats ("old libcpuid", "new libcpiud", "diff-optimized new libcpuid", and "AIDA64") through the `cpuid_deserialize_all_raw_data` function and write only in the "diff-optimized new libcpuid" format, then we don't need to carry over the diff-compression logic to `create_test.py` - we can just add an option to `cpuid_tool` to read a raw dump from one file and write it out to different file; this will do the conversion from AIDA64 to our new format.

> That is too much work to save some kilobytes IMHO.

Yes, don't worry about it, I'll take this task.
Btw, I believe the AIDA64 guys likely would have gone for a diff-optimized format if they knew CPUs will get that many cores back when they devised the format (the '00s?). We have the benefit of hindsight.

> I can improve that and improve cpu_request_core_type at the same time.

Cool! :)

File: libcpuid/cpuid_main.c (hunk 2 above)
Comment (X0rg) [reply libcpuid/cpuid_main.c]: Done in 3a6370474468128918cab560fd26be178e094370.

**Comment by X0rg** [comment]
fb5e29131dcc7be1a11145d54739a34154beeb93: dynamic array of `cpu_raw_data_t` in `cpu_raw_data_array_t`.
fc6b01b074ef87a456ab643904f51800019b299a: opaque `cpu_affinity_mask_t` type (after a deeper look in macros to manage `cpu_set_t` in `sched.h`, I decided to opt to another way).

I still need to work on `cpu_ident_purpose()`.

**Comment by X0rg** [comment]
```
"D:\a\libcpuid\libcpuid\libcpuid_vc10.sln" (default target) (1) ->
"D:\a\libcpuid\libcpuid\libcpuid\libcpuid_vc10.vcxproj" (default target) (2) ->
(Link target) -> 
  cpuid_main.obj : error LNK2001: unresolved external symbol _copy_affinity_mask [D:\a\libcpuid\libcpuid\libcpuid\libcpuid_vc10.vcxproj]
  cpuid_main.obj : error LNK2001: unresolved external symbol _init_affinity_mask [D:\a\libcpuid\libcpuid\libcpuid\libcpuid_vc10.vcxproj]
  cpuid_main.obj : error LNK2001: unresolved external symbol _set_affinity_mask_bit [D:\a\libcpuid\libcpuid\libcpuid\libcpuid_vc10.vcxproj]
  D:\a\libcpuid\libcpuid\libcpuid\x32\ReleaseDLL\libcpuid.dll : fatal error LNK1120: 3 unresolved externals [D:\a\libcpuid\libcpuid\libcpuid\libcpuid_vc10.vcxproj]
```
@anrieff any idea why MSBuild in CI pipeline is not able to find external symbols? `init_affinity_mask` is implemented in `libcpuid_util.c`, header is defined in `libcpuid_util.h` and function is called from `cpuid_main.c`. It builds fine by using CMake on Linux and macOS.
I do not understand what is missing for MSBuild, I compared with existing functions and I do not see any difference.

**Comment by X0rg** [comment]
I improved `cpu_identify_all()` and `cpu_request_core_type()` in 04db409bf48637cd2e098ec93c91cf9e6ab98145.
I found an alternative way to update `num_cores` and `num_logical_cpus` fields for each CPU types with correct values.

I am thinking about a different way to count logical and physical CPUs by using APIC ID. But I discovered this on my system when I use `set_cpu_affinity()`:
```
Identifying logical core 1 with general purpose
APIC ID 0x00000002, package ID 0x00000000, core ID 0x00000001, SMT ID 0x00000002
Identifying logical core 2 with general purpose
APIC ID 0x00000004, package ID 0x00000000, core ID 0x00000002, SMT ID 0x00000004
Identifying logical core 3 with general purpose
APIC ID 0x00000008, package ID 0x00000000, core ID 0x00000004, SMT ID 0x00000008
Identifying logical core 4 with general purpose
APIC ID 0x0000000A, package ID 0x00000000, core ID 0x00000005, SMT ID 0x0000000A
Identifying logical core 5 with general purpose
APIC ID 0x0000000C, package ID 0x00000000, core ID 0x00000006, SMT ID 0x0000000C
Identifying logical core 6 with general purpose
APIC ID 0x00000001, package ID 0x00000000, core ID 0x00000000, SMT ID 0x00000001
Identifying logical core 7 with general purpose
APIC ID 0x00000003, package ID 0x00000000, core ID 0x00000001, SMT ID 0x00000003
Identifying logical core 8 with general purpose
APIC ID 0x00000005, package ID 0x00000000, core ID 0x00000002, SMT ID 0x00000005
Identifying logical core 9 with general purpose
APIC ID 0x00000009, package ID 0x00000000, core ID 0x00000004, SMT ID 0x00000009
Identifying logical core 10 with general purpose
APIC ID 0x0000000B, package ID 0x00000000, core ID 0x00000005, SMT ID 0x0000000B
Identifying logical core 11 with general purpose
APIC ID 0x0000000D, package ID 0x00000000, core ID 0x00000006, SMT ID 0x0000000D
```
APIC ID are not in ascending order... So I cannot use a simple condition like "if previous core ID is different that current core ID, then increment physical CPU count".
Maybe I can use a linked list like the following to properly count physical CPU by purpose based on core ID:
```c
struct internal_apic_linked_list_t {
	logical_cpu_t count;
	int32_t package_id;
	int32_t core_id;
	struct internal_apic_linked_list_t *next;
};
```
The drawback is for the umpteenth core, it will loop until the tail. So, for n physical cores, the formula is (n(n+1))/2. For a system with 2x64 cores, it will loop 8256 times to store all unique CPU ID in the linked list.
I am not sure if it is worth the cost (plus `malloc()` for each new member, etc...).
Maybe I will keep the code "as is" for now.

**Comment by X0rg** [comment]
The `cpu_types` field in `struct system_id_t ` is also a dynamic array since 49de2b56cd26b2a836599c7306fbb1e1708fc325.
I fixed some regressions with serialize/deserialize functions.
I am ready for a new review.

**Comment by anrieff** [comment]
> The drawback is for the umpteenth core, it will loop until the tail. So, for n physical cores, the formula is (n(n+1))/2. For a system with 2x64 cores, it will loop 8256 times to store all unique CPU ID in the linked list.
> I am not sure if it is worth the cost (plus `malloc()` for each new member, etc...).

Well you can fix that by a small hash table.
E.g. make `internal_apic_linked_list_t* htable[HTABLE_SIZE]`, and when considering whether a new entry should be counted or not, traverse the linked list of `htable[entry.core_id % HTABLE_SIZE]`. This will be a small list, typically 0 or 1 elements. Even for a very large CPU, each small linked list would be just 4 items for example.

You can test the hash collision logic (which appends to the small list) by recompiling with intentionally low HTABLE_SIZE, e.g. 4.

As per the other commits - I will review them later today.

**Comment by X0rg** [comment]
> > The drawback is for the umpteenth core, it will loop until the tail. So, for n physical cores, the formula is (n(n+1))/2. For a system with 2x64 cores, it will loop 8256 times to store all unique CPU ID in the linked list.
> > I am not sure if it is worth the cost (plus `malloc()` for each new member, etc...).
> 
> Well you can fix that by a small hash table. E.g. make `internal_apic_linked_list_t* htable[HTABLE_SIZE]`, and when considering whether a new entry should be counted or not, traverse the linked list of `htable[entry.core_id % HTABLE_SIZE]`. This will be a small list, typically 0 or 1 elements. Even for a very large CPU, each small linked list would be just 4 items for example.
> 
> You can test the hash collision logic (which appends to the small list) by recompiling with intentionally low HTABLE_SIZE, e.g. 4.

Ok, for APIC ID, I think I can improve that in #168. For now, with Alder Lake, the way I count logical and physical cores provides correct values.

**Comment by anrieff** [comment]
OK, I think it's good. I've made a few remarks.
I'm not super happy with the way `cpuid_grow_raw_data_array` is called a lot of times (e.g. on a 128-thread machine, 127 `realloc()`s would be called), but let's not worry about it for now. On most systems it would hardly matter at all.
Also, the code implicitly sets a limit of 65536 logical cpus in the system. This is large enough, but we should probably document it more explicitly, in case users wonder what's the limit.

**Review by anrieff — APPROVED** [review]

**Comment by X0rg** [comment]
OK, I took your feedback in account, here is 647e92e8cd2b4b8012e6fa7a39ecd2013a7f7828 (I did not notice before that `cpuid_grow_raw_data_array` and `struct cpu_raw_data_array_t` were limited to 256 logical CPUs).

Windows build still failing due to `unresolved external symbol`. Do you know what is wrong?
And shall we squash this PR before merging?

**Comment by anrieff** [comment]
I think the problem with the unresolved symbols is that e.g. `set_affinity_mask_bit()` is declared `inline` in `libcpuid_util.c`, but not as `extern inline` in `libcpuid_util.h`.
I have no strong preference to merge vs squash.

**Comment by X0rg** [comment]
> I think the problem with the unresolved symbols is that e.g. `set_affinity_mask_bit()` is declared `inline` in `libcpuid_util.c`, but not as `extern inline` in `libcpuid_util.h`. 

Ok, it was that. Thanks!

> I have no strong preference to merge vs squash.

Let's squash then.

**Comment by X0rg** [comment]
I plan to work on other features before releasing a new release (e.g. 0.6.0).
